Add FormattedDatetimeConverter

Daggy1234 opened the pull request:
## Summary

Add datetime converter. This is a draft, hoping for fixes. Uses regex to match discord style timestamps and returns a datetime as well as a flag if present. Draft because at the moment discord has no way of letting regular users send formatted timestamps.

## Checklist

<!-- Put an x inside [ ] to check it, like so: [x] -->

- [x] If code changes were made then they have been tested.
    - [x] I have updated the documentation to reflect the changes.
- [ ] This PR fixes an issue.
- [x] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

Daggy1234 commented:
Extremely sorry for the lack of a description. Clicked on the create button by mistake while adding the description.

JelNiSlaw commented:
I think `re.fullmatch` should be used instead of `re.match`. That would make `$` unnecessary.
Also the rest of the library's code uses `[0-9]` instead of `\d` (I'm not sure why).

JelNiSlaw commented on discord/ext/commands/converter.py:
```diff
@@ -822,6 +824,27 @@ async def convert(self, ctx: Context, argument: str) -> discord.PartialEmoji:
+        if match:
+            try:
+                return datetime.datetime.fromtimestamp(int(match.group(1))), match.group(2)
+            except:
```
You shouldn't be using a bare except

apple502j commented:
> Also the rest of the library's code uses `[0-9]` instead of `\d` (I'm not sure why).

Because:
```py
>>> match(r"^\d+$","٠1１")
<re.Match object; span=(0, 3), match='٠1１'>
```

NCPlayz reviewed (changes requested):
I feel like this would be better if it was a utility function rather than a converter. Something along the lines of `resolve_formatted_timestamp` maybe. It doesn't seem right for the programmer to need to access the converter directly to use as the annotation, and it seems unlikely that a user would be sending a formatted timestamp (over natural language) to specify a date/time argument for a command anyway. 

Just my two cents!

Daggy1234 commented:
So I added the util, but I still kept the converter. The converter also uses the util, and my basic reasoning for retaining the converter was that  eventually discord would add a way for users to send these. Also moved the regex to use [0-9] and added a time zone to the datetime.

Gobot1234 commented on discord/utils.py:
```diff
@@ -1012,3 +1012,29 @@ def format_dt(dt: datetime.datetime, /, style: Optional[TimestampStyle] = None)
+    if match:
+        try:
+            return datetime.datetime.fromtimestamp(int(match.group(1)), tz=datetime.timezone.utc), match.group(2)
+        except:
```
You're using a bare except again

Gobot1234 commented on discord/utils.py:
```diff
@@ -1012,3 +1012,29 @@ def format_dt(dt: datetime.datetime, /, style: Optional[TimestampStyle] = None)
+            return datetime.datetime.fromtimestamp(int(match.group(1)), tz=datetime.timezone.utc), match.group(2)
+        except:
+            return
+    return
```
This is unnecessary

Daggy1234 commented:
In terms of exceptions, for the `int` cast a ValueError, datetime.datetime.fromtimestamp can either raise an OverflowError or a OSError. Both are handled as well. https://docs.python.org/3/library/datetime.html#datetime.date.fromtimestamp

JelNiSlaw commented:
Is it possible for `int` to raise `ValueError` here?

Daggy1234 commented:
> Is it possible for `int` to raise `ValueError` here?

if the regex works probably not. The `int` is for pyright that was complaining a lot.

Rapptz commented:
I'm personally not sure I get the utility of this one. Users aren't necessarily aware of this format at the moment and having to type it manually seems like a chore.

Daggy1234 commented:
Yeah! hence, the utility. Developers can use the utility to parse the timestamps from content bots may send, like editing a timestamp field in an embed. The converter is auxiliary, I can totally remove it, but if discord were to add the ability to send timestamps it exists.
